python/ray/_raylet.pyx

```diff
@@ -1515,11 +1515,18 @@ cdef void async_set_result(shared_ptr[CRayObject] obj,
         data_metadata_pairs, ids_to_deserialize)[0]
 
     def set_future():
+        # Issue #11030, #8841
+        # If this future has result set already, we just need to
+        # skip the set result/exception procedure.
+        if py_future.done():
+            return
+
         if isinstance(result, RayTaskError):
             ray.worker.last_task_error_raise_time = time.time()
             py_future.set_exception(result.as_instanceof_cause())
         else:
             py_future.set_result(result)
+
         cpython.Py_DECREF(py_future)
```

Comment:
We might need to do this for the case you added.

Reply:
Nice catch.